File: app/src/main/java/com/soundcloud/android/users/SocialMediaLinkItem.java
package com.soundcloud.android.users;

import static android.support.v4.content.ContextCompat.getDrawable;

import com.soundcloud.android.api.model.ApiSocialMediaLink;
import com.soundcloud.java.optional.Optional;
import com.soundcloud.java.strings.Strings;

import android.content.Context;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.net.Uri;

public class SocialMediaLinkItem {
    public static SocialMediaLinkItem from(ApiSocialMediaLink apiSocialMediaLink) {
        return create(apiSocialMediaLink.title(), apiSocialMediaLink.network(), apiSocialMediaLink.url());
    }

    public static SocialMediaLinkItem create(Optional<String> title, String network, String url) {
        return new SocialMediaLinkItem(title, Network.from(network), url);
    }

    private final Optional<String> title;
    private final Network network;
    private final String url;

    private SocialMediaLinkItem(Optional<String> title, Network network, String url) {
        this.title = title;
        this.network = network;
        this.url = url;
    }

    public String displayName() {
        return title
                .or(network.displayName())
                .or(url.replaceFirst("^https?://(?:www.)?", Strings.EMPTY));
    }

    public Drawable icon(Context context) {
        return getDrawable(context, network.drawableId());
    }

    public Intent toIntent() {
        if (network.isEmail()) {
            Intent intent = new Intent(Intent.ACTION_SENDTO);
            intent.setData(Uri.parse("mailto:"));
            intent.putExtra(Intent.EXTRA_EMAIL, new String[] { url });
            return intent;
        }
        return new Intent(Intent.ACTION_VIEW, Uri.parse(url));
    }
}

File: app/src/main/java/com/soundcloud/android/users/SocialMediaLinkItem.kt
package com.soundcloud.android.users

import android.content.Context
import android.support.v4.content.ContextCompat.getDrawable
import com.soundcloud.android.api.model.ApiSocialMediaLink
import com.soundcloud.android.utils.OpenForTesting
import com.soundcloud.java.optional.Optional
import com.soundcloud.java.strings.Strings

@OpenForTesting
data class SocialMediaLinkItem(val title: Optional<String>, val network: Network, val url: String) {

    fun displayName(): String {
        return title
                .or(network.displayName())
                .or(url.replaceFirst("^https?://(?:www.)?".toRegex(), Strings.EMPTY))
    }

    fun icon(context: Context) = getDrawable(context, network.drawableId())

    companion object {
        fun from(apiSocialMediaLink: ApiSocialMediaLink) = create(apiSocialMediaLink.title(), apiSocialMediaLink.network(), apiSocialMediaLink.url())

        fun create(title: Optional<String>, network: String, url: String) = SocialMediaLinkItem(title, Network.from(network), url)
    }
}
